test: k8s set [api/base_url] with Helm in Breeze

insomnes opened the pull request:
- Removing setup part in base k8s test class which runs rollout restart on deployment for the kubernetes host setup and related methods
- Set the `[api/base_url]` via parameter from helm chart

closes: #47939 

<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

insomnes committed:
test: k8s set [api/base_url] with Helm in Breeze

Removing setup par in base k8s test class which runs rollout restart on
deployment for the kubernetes host setup

jason810496 commented on dev/breeze/src/airflow_breeze/commands/kubernetes_commands.py:
```diff
@@ -982,6 +982,10 @@ def configure_cluster(
 
 
+CLUSTER_FORWARDED_PORT = os.environ.get("CLUSTER_FORWARDED_PORT") or "8080"
+KUBERNETES_HOST_PORT = (os.environ.get("CLUSTER_HOST") or "localhost") + ":" + CLUSTER_FORWARDED_PORT
```
I think we could add another small utils function for getting `KUBERNETES_HOST_PORT`, or make `_get_kubernetes_port_numbers` public from `kubernetes_utils` module.

insomnes replied:
Sure!

> make _get_kubernetes_port_numbers public from kubernetes_utils module.

Do you want to use it at the breeze scope only? Or it should be used in tests too?

jason810496 replied:
I think the Breeze scope is enough, thanks !

insomnes replied:
I've decided to not go on the `_get_kubernetes_port_numbers` because it has a specific different signature with multiple ports and has different defaults. 

The new helper function `_get_airflow_api_host_with_forwarded_port` encapsulates env vars getting logic. I hope this helps. Or maybe it should be defined in breeze kubernetes utils too?

jason810496 replied:
> I've decided to not go on the _get_kubernetes_port_numbers because it has a specific different signature with multiple ports and has different defaults.  

Since `_deploy_helm_chart` runs with `get_k8s_env` as its environment, and `CLUSTER_FORWARDED_PORT` is passed to the Kubernetes test shell through `new_env` in `get_k8s_env` rather than being set in the current shell,  
I initially thought retrieving `CLUSTER_FORWARDED_PORT` from the current shell might cause errors in getting the **correct** port. However, it turns out the test runs fine.  

That said, I believe fetching it from `_get_kubernetes_port_numbers` might be more accurate.  

> Or maybe it should be defined in breeze kubernetes utils too?  

Yes, the Breeze scope I was referring to is `dev/breeze/src/airflow_breeze/utils/kubernetes_utils.py`. I should have been more specific.

insomnes replied:
> I initially thought retrieving CLUSTER_FORWARDED_PORT from the current shell might cause errors in getting the correct port. However, it turns out the test runs fine.

It's set by breeze within running test helpers, but yeah I can agree with:

>That said, I believe fetching it from _get_kubernetes_port_numbers might be more accurate.

So I used it and also added a warning in case of a non-existing kind config, I didn't change `0, 0` return but found it a bit strange. Maybe there is some case for it during cluster URLs printing for status check, so I left it as is.
I've also set the host in the URL to `localhost` because it's not detected or set anywhere and in all URLs and status checks literal `localhost` is used.

insomnes replied:
I am resolving the comment because you are okay with the implementation now, I believe.

jason810496 reviewed (commented):
Looks good, thanks! Only one small nit.

insomnes committed:
refactor: extract api host + port getting to function

insomnes committed:
test: delete overlooked method _parse_airflow_cfg_dict_as_escaped_toml

insomnes committed:
feat: use get_kubernetes_port_numbers for cluster port detection

- make get_kubernetes_port_numbers "public"
- warning in _get_kind_cluster_config_content in case of non-existing file
- remove previous help function

jason810496 reviewed (approved):
Thanks, the k8s tests work well !

potiuk reviewed (approved):
NICE!
